{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Computer-Aided Vaccine Design\n",
    "\n",
    "In this practical we will rebuild the the epitope selection framework for vaccine design proposed by Toussaint et al.\n",
    "\n",
    "**Remember**: A epitope-based vaccine desing pipeline can be broken down into three essential steps:\n",
    "- Epitope discover\n",
    "- Epitope selection\n",
    "- Epitope Assembly\n",
    "\n",
    "![EV-Design Pipeline](./img/ev_design_pipeline.png)\n",
    "\n",
    "We will cover all three steps in this tutorial. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prerequisites\n",
    "\n",
    "Please make sure you are running a Python2.7 kernel and have Conda installed. You need the following Anaconda packages\n",
    "\n",
    "- Fred2 (A Python Framework for Immunoinformatics)\n",
    "- GLPK (An ILP solver)\n",
    "\n",
    "\n",
    "You can install the packages with the following commands:\n",
    "\n",
    "**Fred2**:\n",
    "pip install git+https://github.com/SchubertLab/Fred2\n",
    "\n",
    "**GLPK**:\n",
    "conda install -c conda-forge glpk\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Epitope Discovery\n",
    "\n",
    "We will design a seasonal Influenza vaccine based on the yearly [WHO reccomendations](https://www.who.int/influenza/vaccines/virus/recommendations/2018_19_north/en/).\n",
    "\n",
    "As targets we will use protein sequences of two recommendet strains (A/Michigan/45/2015 (H1N1) and B/Colorado/06/2017) and will optimize the vaccine's efficacy for the European population.\n",
    "\n",
    "To identify potential epitopes we will use [FRED](https://github.com/SchubertLab/Fred2), a Pyhon-based API for immunoinformatics that interfaces with many epitope affinity prediction methods. \n",
    "\n",
    "You might find FREDs [documentation](https://fred2.readthedocs.io/en/latest/) and [tutorial](https://github.com/FRED-2/Fred2/tree/master/Fred2/tutorials) usefull.\n",
    "\n",
    "Task 1.1: Read in the data and use FRED to generate Peptides and HLA objects\n",
    "----\n",
    "Use FRED's functions to read in the fasta file and HLA allele file to generate Peptide and HLA objects. Please filter the HLA alleles for a frequency of at least 5%\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Fred2.Core import Allele, Peptide, Protein,generate_peptides_from_proteins\n",
    "from Fred2.IO import read_lines, read_fasta\n",
    "\n",
    "proteins = read_fasta(\"./data/Influenza_ B_Colorado_06_2017_ A_Michigan_45_2015.fasta\", in_type=Protein, id_position=0)\n",
    "peptides = list(generate_peptides_from_proteins(proteins, window_size=9))\n",
    "\n",
    "\n",
    "alleles = []\n",
    "with open(\"./data/HLA_alleles_europe.tsv\", \"r\") as f:\n",
    "    for l in f:\n",
    "        allele,prob = l.split()\n",
    "        if float(prob) >= 0.05:\n",
    "            alleles.append(Allele(allele,prob=float(prob.strip())))\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task 1.2: Epitope prediction\n",
    "----\n",
    "Then, use an epitope prediction method to identify potential candidate epitopes for vaccination and filter epitopes whit a binding affinity of at most 50 nM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "from Fred2.EpitopePrediction import EpitopePredictorFactory\n",
    "#you can either use pre-defined operators from `operator`\n",
    "from operator import ge\n",
    "#or define you own comparator function like this\n",
    "comparator = lambda a,b: a <= b\n",
    "\n",
    "smmpmbc = EpitopePredictorFactory(\"smmpmbec\")\n",
    "\n",
    "ep_df = smmpmbc.predict(peptides, alleles=alleles)\n",
    "\n",
    "ep_df_filtered = ep_df.filter_result([(\"smmpmbec\",comparator, 50)])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Epitope Selection\n",
    "\n",
    "To select the optimal set of epitopes for vaccination, we will rebuild OptiTope, Toussaints et al.'s epitope selection ILP, using Pyomo. Pyomo ist a Python-based modeling language for constraint optimization problems. We will sucessively build ILP formulation and in the ende will use it to select the optimal set of epitopes.\n",
    "\n",
    "### 2.1 Overview\n",
    "\n",
    "ILP formulations have five essential element: Sets, Parameters, Variables, Objectives, and Constraints. The letter two use sets as indices for parameters and variables to formulate the equations that make up the objectives and constraints.\n",
    "\n",
    "\n",
    "As a reminder, the complete formulation of OptiTope looks like this:\n",
    "\n",
    "\n",
    "\\begin{align}\n",
    "   \\text{Sets:}&&&\\\\\n",
    "   \\\\\n",
    "    E&&&\\text{Set of epitopes}\\\\\n",
    "    H&&&\\text{Set of HLA alleles}\\\\\n",
    "    A&&&\\text{Set of antigens}\\\\\n",
    "    I(h)\\subseteq E&~\\forall~h \\in H&&\\text{Set of epitopes binding to HLA $h$}\\\\\n",
    "    \\\\\n",
    "    \\text{Parameters:}&&&\\\\\n",
    "    \\\\\n",
    "    i_{e, h}&&&\\text{Immunogenicity of epitope $e$ bound to HLA $h$}\\\\\n",
    "    p_h&&&\\text{Allele frequency of HLA $h$ in target population}\\\\\n",
    "    k&&&\\text{Upper limit of selected epitopes}\\\\\n",
    "    \\tau_{H}&&&\\text{Minimum of covered HLA alleles}\\\\\n",
    "    \\\\\n",
    "    \\text{Variables:}&&&\\\\\n",
    "    \\\\\n",
    "    x_e \\in \\{0,1\\}&~\\forall~e \\in E&&\\text{Dicision variable whether epitope $e$ is selected}\\\\\n",
    "    y_h\\in \\{0,1\\}&~\\forall~h \\in H&&\\text{Dicision variable whether HLA $h$ is covered by the selected epitopes}\\\\\n",
    "    \\\\\n",
    "    \\text{Objective:}&&&\\\\\n",
    "    \\\\\n",
    "    \\arg\\max_{\\mathbf{x}} &\\sum_{h \\in H} p_h \\sum_{e \\in E} x_ei_{e,h}&&\\\\\n",
    "    \\\\\n",
    "    \\text{Constraints:}&&&\\\\\n",
    "    \\\\\n",
    "    \\sum_{e\\in E} x_e \\leq k&&&\\\\\n",
    "    \\sum_{e\\in I(h)} x_e \\geq y_h &~\\forall h \\in H&&\\\\\n",
    "    \\sum_{h \\in H} y_h \\leq \\tau_{H} &~\\forall h \\in H&&\\\\\n",
    "\\end{align}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Abstract Versus Concrete Models\n",
    "\n",
    "*(Adopted from Pyomo Documentation)*\n",
    "\n",
    "A mathematical model can be defined using symbols that represent data\n",
    "values.  For example, the following equations represent a linear program\n",
    "(LP) to find optimal values for the vector $x$ with parameters\n",
    "$n$ and $b$, and parameter vectors $a$ and $c$:\n",
    "\n",
    "\n",
    "\\begin{array}{lll}\n",
    "    \\min       & \\sum_{j=1}^n c_j x_j &\\\\\n",
    "     \\mathrm{s.t.} & \\sum_{j=1}^n a_{ij} x_j \\geq b_i & \\forall i = 1 \\ldots m\\\\\n",
    "               & x_j \\geq 0 & \\forall j = 1 \\ldots n\n",
    "\\end{array}\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "   As a convenience, we use the symbol $\\forall$ to mean \"for all\"\n",
    "   or \"for each.\"\n",
    "</div>\n",
    "\n",
    "We call this an *abstract* or *symbolic* mathematical model since it\n",
    "relies on unspecified parameter values.  Data values can be used to\n",
    "specify a *model instance*.  The ``AbstractModel`` class provides a\n",
    "context for defining and initializing abstract optimization models in\n",
    "Pyomo when the data values will be supplied at the time a solution is to\n",
    "be obtained.\n",
    "\n",
    "In many contexts, a mathematical model can and should be directly\n",
    "defined with the data values supplied at the time of the model\n",
    "definition.  We call these *concrete* mathematical models.  For example,\n",
    "the following LP model is a concrete instance of the previous abstract\n",
    "model:\n",
    "\n",
    "\n",
    "\\begin{array}{ll}\n",
    "    \\min       & 2 x_1 + 3 x_2\\\\\n",
    "     \\mathrm{s.t.} & 3 x_1 + 4 x_2 \\geq 1\\\\\n",
    "               & x_1, x_2 \\geq 0\n",
    "\\end{array}\n",
    "\n",
    "The ``ConcreteModel`` class is used to define concrete optimization\n",
    "models in Pyomo.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "   Python programmers will probably prefer to write concrete models,\n",
    "   while users of some other algebraic modeling languages may tend to\n",
    "   prefer to write abstract models.  The choice is largely a matter of\n",
    "   taste; some applications may be a little more straightforward using\n",
    "   one or the other.\n",
    "</div>\n",
    "\n",
    "\n",
    "We will work with a ``ConcreteModel``. You can find an example of a ``ConcreteModel`` [here](https://pyomo.readthedocs.io/en/latest/pyomo_overview/simple_examples.html#a-simple-concrete-pyomo-model).\n",
    "\n",
    "Task 2.1: ConcreteModel initialization\n",
    "-----\n",
    "1) Import the necessary Pyomo modules and initialize a concrete model.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyomo.environ import *\n",
    "\n",
    "model = ConcreteModel()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Set and Sets of Sets\n",
    "*(Adopted from Pyomo Documentation)*\n",
    "\n",
    "\n",
    "Sets can be declared using the `Set` and `RangeSet` functions or by\n",
    "assigning set expressions.  The simplest set declaration creates a set\n",
    "and postpones creation of its members:\n",
    "\n",
    "`\n",
    "model.A = Set()\n",
    "`\n",
    "\n",
    "The ``Set`` function takes optional arguments such as:\n",
    "\n",
    "- doc = String describing the set\n",
    "- dimen = Dimension of the members of the set\n",
    "- filter = A boolean function used during construction to indicate if a\n",
    "  potential new member should be assigned to the set\n",
    "- initialize = A function that returns the members to initialize the set.\n",
    "- ordered = A boolean indicator that the set is ordered; the default is `False`\n",
    "- validate = A boolean function that validates new member data\n",
    "- virtual = A boolean indicator that the set will never have elements;\n",
    "  it is unusual for a modeler to create a virtual set; they are\n",
    "  typically used as domains for sets, parameters and variables\n",
    "- within = Set used for validation; it is a super-set of the set being declared.\n",
    "\n",
    "\n",
    "The `initialize` option can refer to a Python set, which can be\n",
    "returned by a function or given directly as in\n",
    "\n",
    "`\n",
    "model.D = Set(initialize=['red', 'green', 'blue'])\n",
    "`\n",
    "\n",
    "If sets are given as arguments to `Set` without keywords, they are\n",
    "interpreted as indexes for an array of sets. For example, to create an\n",
    "array of sets that is indexed by the members of the set `model.A`, use\n",
    "\n",
    "`\n",
    "model.E = Set(model.A)\n",
    "`\n",
    "\n",
    "This is not restricted to one-dimentional indices. You can specify abitrary dimentional index structures.\n",
    "\n",
    "You can use `initialize` to construct a concrete array of sets either by providing a function that takes the model and an element of `model.A`as input, and returns the entire set, or by providing a dictionary where the `key` is the an element of `model.A` and the `value` the entire set.  \n",
    "\n",
    "\n",
    "You can finde more detail [here](https://pyomo.readthedocs.io/en/latest/pyomo_overview/simple_examples.html#a-simple-concrete-pyomo-model)\n",
    "\n",
    "\n",
    "Task 2.2: OptiType's Set definition\n",
    "---\n",
    "Define and initialize the Sets of OptiType with the predicted epitope data.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "\n",
    "<b>Note:</b> The column and row indices of the epitope prediction dataframe contain `Peptide` and `HLA-Allele` Objects.\n",
    "Through the functionality of the Objects you have access to the Protein origine of the Peptides and the allele probability in the target population.\n",
    "\n",
    "You can find more information about the `Peptide`-Object [here](https://fred2.readthedocs.io/en/latest/Fred2.Core.html#module-Fred2.Core.Peptide) and for the `Allele`-Object [here](https://fred2.readthedocs.io/en/latest/Fred2.Core.html#module-Fred2.Core.Allele).\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "eps = ep_df_filtered.index.get_level_values(0).tolist()\n",
    "alleles = ep_df_filtered.columns.tolist()\n",
    "\n",
    "#Simple Sets\n",
    "model.E = Set(initialize=map(str, eps))\n",
    "model.H = Set(initialize=map(str, alleles))\n",
    "\n",
    "# Array of Sets\n",
    "def init_I(model, h):\n",
    "    d = ep_df_filtered.loc[:, h]\n",
    "    d2 = d[d <= 50]\n",
    "    return d2.index.get_level_values(0).tolist()        \n",
    "model.I = Set(model.H, initialize=init_I)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4 Parameters \n",
    "*(Adopted from Poymo Documentation)*\n",
    "\n",
    "The word “parameters” is used in many settings. When discussing a Pyomo model, we use the word to refer to data that must be provided in order to find an optimal (or good) assignment of values to the decision variables. Parameters are declared with the `Param` function, which takes arguments that are somewhat similar to the Set function. For example, the following code snippet declares sets `model.A`, `model.B` and then a parameter array `model.P` that is indexed by `model.A`:\n",
    "\n",
    "`\n",
    "model.A = Set()\n",
    "model.B = Set()\n",
    "model.P = Param(model.A, model.B)\n",
    "`\n",
    "\n",
    "In addition to sets that serve as indexes, the `Param` function takes the following command options:\n",
    "\n",
    "- default = The value absent any other specification.\n",
    "- doc = String describing the parameter\n",
    "- initialize = A function (or Python object) that returns the members to initialize the parameter values.\n",
    "- validate = A boolean function with arguments that are the prospective parameter value, the parameter indices and the model.\n",
    "- within = Set used for validation; it specifies the domain of the parameter values.\n",
    "\n",
    "These options perform in the same way as they do for Set. For example, suppose that `model.A = RangeSet(1,3)`, then there are many ways to create a parameter that is a square matrix with 9, 16, 25 on the main diagonal zeros elsewhere, here are two ways to do it. First using a Python object to initialize:\n",
    "\n",
    "`\n",
    "v={}\n",
    "v[1,1] = 9\n",
    "v[2,2] = 16\n",
    "v[3,3] = 25\n",
    "model.S = Param(model.A, model.A, initialize=v, default=0)\n",
    "`\n",
    "\n",
    "And now using an initialization function that is automatically called once for each index tuple (remember that we are assuming that `model.A` contains 1,2,3)\n",
    "\n",
    "`\n",
    "def s_init(model, i, j):\n",
    "    if i == j:\n",
    "        return i*i\n",
    "    else:\n",
    "        return 0.0\n",
    "model.S1 = Param(model.A, model.A, initialize=s_init)\n",
    "`\n",
    "\n",
    "In this example, the index set contained integers, but index sets need not be numeric. It is very common to use strings."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task 2.3: OptiType's Parameter definition\n",
    "---\n",
    "\n",
    "Define and initialize the Paremters of OptiType.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "\n",
    "model.k = Param(initialize=10, within=PositiveIntegers, mutable=True)\n",
    "\n",
    "probs = {str(h):h.prob for h in alleles}\n",
    "model.p = Param(model.H, initialize=lambda model, a: probs[a])\n",
    "model.t_allele = Param(initialize=0, within=NonNegativeIntegers, mutable=True)\n",
    "model.i = Param(model.E, model.H, initialize=lambda model, e, h: min(1., max(0.0, 1.0 - math.log(ep_df_filtered.loc[(e, \"smmpmbec\"), h], 50000))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.5 Variables\n",
    "*(Adopted from Pyomo Documentation)*\n",
    "\n",
    "Variables are intended to ultimately be given values by an optimization package. They are declared and optionally bounded, given initial values, and documented using the Pyomo `Var` function. If index sets are given as arguments to this function they are used to index the variable. Other optional directives include:\n",
    "\n",
    "- bounds = A function (or Python object) that gives a (lower,upper) bound pair for the variable\n",
    "- domain = A set that is a super-set of the values the variable can take on.\n",
    "- initialize = A function (or Python object) that gives a starting value for the variable; this is particularly important for non-linear models\n",
    "- within = (synonym for domain)\n",
    "\n",
    "The following code snippet illustrates some aspects of these options by declaring a singleton (i.e. unindexed) variable named `model.LumberJack` that will take on real values between zero and 6 and it initialized to be 1.5:\n",
    "\n",
    "`\n",
    "model.LumberJack = Var(within=NonNegativeReals, bounds=(0,6), initialize=1.5)\n",
    "`\n",
    "\n",
    "\n",
    "For more details see [here](https://pyomo.readthedocs.io/en/latest/pyomo_modeling_components/Variables.html).\n",
    "\n",
    "\n",
    "Task 2.4: OptiTope's Variable definition\n",
    "---\n",
    "Define OptiTope's variables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.x = Var(model.E, within=Binary)\n",
    "model.y = Var(model.H, within=Binary)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.6 Objectives\n",
    "*(Adopted from Pyomos Documentation)*\n",
    "\n",
    "An objective is a function of variables that returns a value that an optimization package attempts to maximize or minimize. The `Objective` function in Pyomo declares an objective. Although other mechanisms are possible, this function is typically passed the name of another function that gives the expression. Here is a very simple version of such a function that assumes `model.x` has previously been declared as a `Var`:\n",
    "\n",
    "\n",
    "    def ObjRule(model):\n",
    "        return 2*model.x[1] + 3*model.x[2]\n",
    "    model.g = Objective(rule=ObjRule)\n",
    "\n",
    "\n",
    "It is more common for an objective function to refer to parameters as in this example that assumes that `model.p` has been declared as a `Param` and that `model.x` has been declared with the same index set, while `model.y` has been declared as a singleton:\n",
    "\n",
    "\n",
    "    def profrul(model):\n",
    "        return summation(model.p, model.x) + model.y\n",
    "    model.Obj = Objective(rule=ObjRule, sense=maximize)\n",
    "\n",
    "\n",
    "This example uses the `sense` option to specify maximization. The default `sense` is minimize.\n",
    "\n",
    "Task 2.5: OptiTope's Objective definition\n",
    "---\n",
    "Define OptiTope's objetive function\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.obj = Objective(\n",
    "            rule=lambda model: sum(model.x[e] * sum(model.p[a] * model.i[e, a] for a in model.H) for e in model.E),\n",
    "            sense=maximize)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.7 Constraints\n",
    "*(Adopted from Pyomos Documentation)*\n",
    "\n",
    "Most constraints are specified using equality or inequality expressions that are created using a rule, which is a Python function. For example, if the variable `model.x` has the indexes ‘butter’ and ‘scones’, then this constraint limits the sum over these indexes to be exactly three:\n",
    "\n",
    "    def teaOKrule(model):\n",
    "        return(model.x['butter'] + model.x['scones'] == 3)\n",
    "    model.TeaConst = Constraint(rule=teaOKrule)\n",
    "\n",
    "Instead of expressions involving equality (==) or inequalities (<= or >=), constraints can also be expressed using a 3-tuple if the form (lb, expr, ub) where lb and ub can be None, which is interpreted as lb <= expr <= ub. Variables can appear only in the middle expr. For example, the following two constraint declarations have the same meaning:\n",
    "\n",
    "    model.x = Var()\n",
    "\n",
    "    def aRule(model):\n",
    "       return model.x >= 2\n",
    "    model.Boundx = Constraint(rule=aRule)\n",
    "\n",
    "    def bRule(model):\n",
    "       return (2, model.x, None)\n",
    "    model.boundx = Constraint(rule=bRule)\n",
    "\n",
    "Constraints (and objectives) can be indexed by lists or sets. When the declaration contains lists or sets as arguments, the elements are iteratively passed to the rule function. If there is more than one, then the cross product is sent. For example the following constraint could be interpreted as placing a budget of `i` on the ith item to buy where the cost per item is given by the parameter `model.a`:\n",
    "\n",
    "    model.A = RangeSet(1,10)\n",
    "    model.a = Param(model.A, within=PositiveReals)\n",
    "    model.ToBuy = Var(model.A)\n",
    "    def bud_rule(model, i):\n",
    "        return model.a[i]*model.ToBuy[i] <= i\n",
    "    aBudget = Constraint(model.A, rule=bud_rule)\n",
    "\n",
    "Task 2.6: OptiTope's Constraints\n",
    "---\n",
    "Define OptiTope's constraints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Obligatory Constraint (number of selected epitopes)\n",
    "model.NofSelectedEpitopesCov = Constraint(rule=lambda model: sum(model.x[e] for e in model.E) <= model.k)\n",
    "\n",
    "#optional constraints (in basic model they are disabled)\n",
    "model.IsAlleleCovConst = Constraint(model.H,\n",
    "                                    rule=lambda model, h: sum(model.x[e] for e in model.I[h]) >= model.y[h])\n",
    "model.MinAlleleCovConst = Constraint(rule=lambda model: sum(model.y[h] for a in model.H) >= model.t_allele)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.8 Solving a `ConcreteModels`\n",
    "\n",
    "If you have a ConcreteModel, add these lines at the bottom of your Python script to solve it\n",
    "\n",
    "    opt = pyo.SolverFactory('glpk')\n",
    "    res = opt.solve(model) \n",
    "    \n",
    "    \n",
    "To load the optimal parameters into your model you can use this snippet:\n",
    "\n",
    "    model.solutions.load_from(res)\n",
    "    selected_peptides = [x for x in self.instance.x if 0.9 <= model.x[x].value <= 1.2]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# ==========================================================\n",
      "# = Solver Results                                         =\n",
      "# ==========================================================\n",
      "# ----------------------------------------------------------\n",
      "#   Problem Information\n",
      "# ----------------------------------------------------------\n",
      "Problem: \n",
      "- Name: unknown\n",
      "  Lower bound: 5.61031116551\n",
      "  Upper bound: 5.61031116551\n",
      "  Number of objectives: 1\n",
      "  Number of constraints: 17\n",
      "  Number of variables: 2009\n",
      "  Number of nonzeros: 4154\n",
      "  Sense: maximize\n",
      "# ----------------------------------------------------------\n",
      "#   Solver Information\n",
      "# ----------------------------------------------------------\n",
      "Solver: \n",
      "- Status: ok\n",
      "  Termination condition: optimal\n",
      "  Statistics: \n",
      "    Branch and bound: \n",
      "      Number of bounded subproblems: 1\n",
      "      Number of created subproblems: 1\n",
      "  Error rc: 0\n",
      "  Time: 0.0518410205841\n",
      "# ----------------------------------------------------------\n",
      "#   Solution Information\n",
      "# ----------------------------------------------------------\n",
      "Solution: \n",
      "- number of solutions: 0\n",
      "  number of solutions displayed: 0\n"
     ]
    }
   ],
   "source": [
    "from pyomo.opt import SolverFactory, TerminationCondition\n",
    "\n",
    "opt = SolverFactory('glpk')\n",
    "res = opt.solve(model) \n",
    "\n",
    "res.write(num=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MMMGMFNML',\n",
       " 'LLMDGTASL',\n",
       " 'YLMAWKQVL',\n",
       " 'FMQALQLLL',\n",
       " 'YVLFHTSLL',\n",
       " 'YVSASLSYL',\n",
       " 'YLNPGNYSM',\n",
       " 'FMYSDFHFI',\n",
       " 'FVANFSMEL',\n",
       " 'KMMKVLLFM']"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "model.solutions.load_from(res)\n",
    "selected_peptides = [x for x in model.x if 0.9 <= model.x[x].value <= 1.2]\n",
    "selected_peptides\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python [conda env:python2.7]",
   "language": "python",
   "name": "conda-env-python2.7-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1.0,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1.0,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
