{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computer-Aided Vaccine Design\n",
    "\n",
    "In this practical we will implement a neoeptiope selection framework  based on Toussaint et al methematical model.\n",
    "\n",
    "This will include:\n",
    "\n",
    "1) Neopeptide construction from somatic mutations\n",
    "2) Neoeptioep prediction\n",
    "3) Neoepitope selection under constraints\n",
    "4) Vaccine assembly\n",
    "\n",
    "We will cover all three steps in this tutorial. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prerequisites\n",
    "\n",
    "Please clone the repository and install the following packages environment \n",
    "- PySAM (Python API to read and wirte SAM/BAM/VCF/BCF)\n",
    "- Pyomo (Python API for Integer linear programming)\n",
    "- Epytope (A Python Framework for Immunoinformatics)\n",
    "- GLPK (An ILP solver\n",
    "\n",
    "or use the pre-existing conda:\n",
    "```\n",
    "pip uninstall pyomo\n",
    "pip install pyomo\n",
    "```\n",
    "\n",
    "You can install the packages with the following commands:\n",
    "\n",
    "**GLPK**:\n",
    "!apt-get install glpk-utils\n",
    "\n",
    "**Epytope**:\n",
    "pip install git+https://github.com/KohlbacherLab/epytope\n",
    "\n",
    "\n",
    "\n",
    "Than please uninstall pyomo and reinstall it again:\n",
    "\n",
    "\n",
    "**Pyomo**\n",
    "```\n",
    "pip uninstall pyomo\n",
    "pip install pyomo\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Neoepitope construction\n",
    "\n",
    "We will start we implementing functions to read VEP files (output format of the ENSEMBLES Variant effect Predictor, which we used to annotate identified somatic mutations using MuTec2, and create alterated peptides form the variances.\n",
    "You can finde the exmaple file under ```./data/VEP_example.vep```.\n",
    "\n",
    "To this end, we need:\n",
    "\n",
    "1) Collect nonsynonymous single point variances for a transcript \n",
    "2) Extract the reference transcript sequence \n",
    "3) Integrate the somatic mutations\n",
    "4) Tranlsate the RNA sequence into its amino acid sequence\n",
    "5) Create all possible k-mers (while keeping track to which protein it belongt); k in [9, 11]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task 1.1: Implement a VEP reader and a representations of ```Variants```\n",
    "----------\n",
    "\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task 1.2: Implement a neopeptide generator function that consumes a list of ```Variants``` an generates a list of ```Peptides```of length k\n",
    "----------\n",
    "Assume that the variants were created usign the human reference genome GRCh38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2: Neoepitope prediction\n",
    "\n",
    "Once we have created all possible possible alterated neopeptides we need to predict which can bind to HLA and subsequently will be presented on the surface of the tumor.\n",
    "\n",
    "To this end, we will use MHCflurry. MHCflurry 2.0 is a Deep Learning model using feed-forward neural networks with skip-connections. If you want to learn more about the model check out their paper [here](https://www.cell.com/cell-systems/fulltext/S2405-4712(20)30239-8?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS2405471220302398%3Fshowall%3Dtrue).\n",
    "\n",
    "You cand find the github repository [here](https://github.com/openvax/mhcflurry) and an example of its API [here](https://github.com/openvax/mhcflurry/blob/master/notebooks/example1.ipynb).\n",
    "\n",
    "![img](./img/mhcflurry.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task 2.1: Implement a Neoepitope function who consums a list of generated ```Peptides``` and predicted their binding affinity to a list of ```HLAs```\n",
    "-----\n",
    "Assume the patients HLA-I genotype is HLA-A\\*02:01, HLA-B\\*07:02, and HLA-C\\*02:02"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Epitope Selection\n",
    "\n",
    "To select the optimal set of epitopes for vaccination we will rebuild OptiTope, Toussaints et al.'s epitope selection ILP, using Pyomo. Pyomo ist a Python-based modeling language for constraint optimization problems. Besides implementing the ILP model, your job will be to come up with a Objective function that is specific for neoeptiope selection and the document & justification your formulation.\n",
    "\n",
    "### 3.1 Overview\n",
    "\n",
    "ILP formulations have five essential element: `Sets`, `Parameters`, `Variables`, `Objectives`, and `Constraints`. The letter two use sets as indices for parameters and variables to formulate the equations that make up the objectives and constraints.\n",
    "\n",
    "\n",
    "As a reminder, the complete formulation of OptiTope looks like this:\n",
    "\n",
    "\n",
    "\\begin{align}\n",
    "   \\text{Sets:}&&&\\\\\n",
    "   \\\\\n",
    "    E&&&\\text{Set of epitopes}\\\\\n",
    "    H&&&\\text{Set of HLA alleles}\\\\\n",
    "    I(h)\\subseteq E&~\\forall~h \\in H&&\\text{Set of epitopes binding to HLA $h$}\\\\\n",
    "    \\\\\n",
    "    \\text{Parameters:}&&&\\\\\n",
    "    \\\\\n",
    "    i_{e, h}&&&\\text{Immunogenicity of epitope $e$ bound to HLA $h$}\\\\\n",
    "    p_h&&&\\text{Allele frequency of HLA $h$ in target population}\\\\\n",
    "    k&&&\\text{Upper limit of selected epitopes}\\\\\n",
    "    \\tau_{H}&&&\\text{Minimum of covered HLA alleles}\\\\\n",
    "    \\\\\n",
    "    \\text{Variables:}&&&\\\\\n",
    "    \\\\\n",
    "    x_e \\in \\{0,1\\}&~\\forall~e \\in E&&\\text{Dicision variable whether epitope $e$ is selected}\\\\\n",
    "    y_h\\in \\{0,1\\}&~\\forall~h \\in H&&\\text{Dicision variable whether HLA $h$ is covered by the selected epitopes}\\\\\n",
    "    \\\\\n",
    "    \\text{Objective:}&&&\\\\\n",
    "    \\\\\n",
    "    \\arg\\max_{\\mathbf{x}} &\\sum_{h \\in H} p_h \\sum_{e \\in E} x_ei_{e,h}&&\\\\\n",
    "    \\\\\n",
    "    \\text{Constraints:}&&&\\\\\n",
    "    \\\\\n",
    "    \\sum_{e\\in E} x_e \\leq k&&&\\\\\n",
    "    \\sum_{e\\in I(h)} x_e \\geq y_h &~\\forall h \\in H&&\\\\\n",
    "    \\sum_{h \\in H} y_h \\leq \\tau_{H} &~\\forall h \\in H&&\\\\\n",
    "\\end{align}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Abstract Versus Concrete Models\n",
    "\n",
    "*(Adopted from Pyomo Documentation)*\n",
    "\n",
    "A mathematical model can be defined using symbols that represent data\n",
    "values.  For example, the following equations represent a linear program\n",
    "(LP) to find optimal values for the vector $x$ with parameters\n",
    "$n$ and $b$, and parameter vectors $a$ and $c$:\n",
    "\n",
    "\n",
    "\\begin{array}{lll}\n",
    "    \\min       & \\sum_{j=1}^n c_j x_j &\\\\\n",
    "     \\mathrm{s.t.} & \\sum_{j=1}^n a_{ij} x_j \\geq b_i & \\forall i = 1 \\ldots m\\\\\n",
    "               & x_j \\geq 0 & \\forall j = 1 \\ldots n\n",
    "\\end{array}\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "   As a convenience, we use the symbol $\\forall$ to mean \"for all\"\n",
    "   or \"for each.\"\n",
    "</div>\n",
    "\n",
    "We call this an *abstract* or *symbolic* mathematical model since it\n",
    "relies on unspecified parameter values.  Data values can be used to\n",
    "specify a *model instance*.  The ``AbstractModel`` class provides a\n",
    "context for defining and initializing abstract optimization models in\n",
    "Pyomo when the data values will be supplied at the time a solution is to\n",
    "be obtained.\n",
    "\n",
    "In many contexts, a mathematical model can and should be directly\n",
    "defined with the data values supplied at the time of the model\n",
    "definition.  We call these *concrete* mathematical models.  For example,\n",
    "the following LP model is a concrete instance of the previous abstract\n",
    "model:\n",
    "\n",
    "\n",
    "\\begin{array}{ll}\n",
    "    \\min       & 2 x_1 + 3 x_2\\\\\n",
    "     \\mathrm{s.t.} & 3 x_1 + 4 x_2 \\geq 1\\\\\n",
    "               & x_1, x_2 \\geq 0\n",
    "\\end{array}\n",
    "\n",
    "The ``ConcreteModel`` class is used to define concrete optimization\n",
    "models in Pyomo.\n",
    "\n",
    "<div class=\"alert alert-block alert-info\">\n",
    "   Python programmers will probably prefer to write concrete models,\n",
    "   while users of some other algebraic modeling languages may tend to\n",
    "   prefer to write abstract models.  The choice is largely a matter of\n",
    "   taste; some applications may be a little more straightforward using\n",
    "   one or the other.\n",
    "</div>\n",
    "\n",
    "\n",
    "We will work with a ``ConcreteModel``. You can find an example of a ``ConcreteModel`` [here](https://pyomo.readthedocs.io/en/latest/pyomo_overview/simple_examples.html#a-simple-concrete-pyomo-model).\n",
    "\n",
    "Task 3.1: ConcreteModel initialization\n",
    "----------\n",
    "1) Import the necessary Pyomo modules and initialize a concrete model.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Set and Sets of Sets\n",
    "*(Adopted from Pyomo Documentation)*\n",
    "\n",
    "\n",
    "Sets can be declared using the `Set` and `RangeSet` functions or by\n",
    "assigning set expressions.  The simplest set declaration creates a set\n",
    "and postpones creation of its members:\n",
    "\n",
    "`\n",
    "model.A = Set()\n",
    "`\n",
    "\n",
    "The ``Set`` function takes optional arguments such as:\n",
    "\n",
    "- doc = String describing the set\n",
    "- dimen = Dimension of the members of the set\n",
    "- filter = A boolean function used during construction to indicate if a\n",
    "  potential new member should be assigned to the set\n",
    "- initialize = A function that returns the members to initialize the set.\n",
    "- ordered = A boolean indicator that the set is ordered; the default is `False`\n",
    "- validate = A boolean function that validates new member data\n",
    "- virtual = A boolean indicator that the set will never have elements;\n",
    "  it is unusual for a modeler to create a virtual set; they are\n",
    "  typically used as domains for sets, parameters and variables\n",
    "- within = Set used for validation; it is a super-set of the set being declared.\n",
    "\n",
    "\n",
    "The `initialize` option can refer to a Python set, which can be\n",
    "returned by a function or given directly as in\n",
    "\n",
    "`\n",
    "model.D = Set(initialize=['red', 'green', 'blue'])\n",
    "`\n",
    "\n",
    "If sets are given as arguments to `Set` without keywords, they are\n",
    "interpreted as indexes for an array of sets. For example, to create an\n",
    "array of sets that is indexed by the members of the set `model.A`, use\n",
    "\n",
    "`\n",
    "model.E = Set(model.A)\n",
    "`\n",
    "\n",
    "This is not restricted to one-dimentional indices. You can specify abitrary dimentional index structures.\n",
    "\n",
    "You can use `initialize` to construct a concrete array of sets either by providing a function that takes the model and an element of `model.A`as input, and returns the entire set, or by providing a dictionary where the `key` is the an element of `model.A` and the `value` the entire set.  \n",
    "\n",
    "\n",
    "You can finde more detail [here](https://pyomo.readthedocs.io/en/latest/pyomo_overview/simple_examples.html#a-simple-concrete-pyomo-model)\n",
    "\n",
    "\n",
    "Task 3.2: OptiTope's Set definition\n",
    "---\n",
    "Define and initialize the Sets of OptiTope with the predicted epitope data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4 Parameters \n",
    "*(Adopted from Poymo Documentation)*\n",
    "\n",
    "The word “parameters” is used in many settings. When discussing a Pyomo model, we use the word to refer to data that must be provided in order to find an optimal (or good) assignment of values to the decision variables. Parameters are declared with the `Param` function, which takes arguments that are somewhat similar to the Set function. For example, the following code snippet declares sets `model.A`, `model.B` and then a parameter array `model.P` that is indexed by `model.A`:\n",
    "\n",
    "```\n",
    "model.A = Set()\n",
    "model.B = Set()\n",
    "model.P = Param(model.A, model.B)\n",
    "```\n",
    "\n",
    "In addition to sets that serve as indexes, the `Param` function takes the following command options:\n",
    "\n",
    "- default = The value absent any other specification.\n",
    "- doc = String describing the parameter\n",
    "- initialize = A function (or Python object) that returns the members to initialize the parameter values.\n",
    "- validate = A boolean function with arguments that are the prospective parameter value, the parameter indices and the model.\n",
    "- within = Set used for validation; it specifies the domain of the parameter values.\n",
    "\n",
    "These options perform in the same way as they do for Set. For example, suppose that `model.A = RangeSet(1,3)`, then there are many ways to create a parameter that is a square matrix with 9, 16, 25 on the main diagonal zeros elsewhere, here are two ways to do it. First using a Python object to initialize:\n",
    "\n",
    "```\n",
    "v={}\n",
    "v[1,1] = 9\n",
    "v[2,2] = 16\n",
    "v[3,3] = 25\n",
    "model.S = Param(model.A, model.A, initialize=v, default=0)\n",
    "```\n",
    "\n",
    "And now using an initialization function that is automatically called once for each index tuple (remember that we are assuming that `model.A` contains 1,2,3)\n",
    "\n",
    "```\n",
    "def s_init(model, i, j):\n",
    "    if i == j:\n",
    "        return i*i\n",
    "    else:\n",
    "        return 0.0\n",
    "model.S1 = Param(model.A, model.A, initialize=s_init)\n",
    "```\n",
    "\n",
    "In this example, the index set contained integers, but index sets need not be numeric. It is very common to use strings."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task 3.3: OptiTope's Parameter definition\n",
    "---\n",
    "\n",
    "Define and initialize the Paremters of OptiTope. Since the predicted binding affinities are IC50 values, we have to transform the data with the following functio:\n",
    "\n",
    "\\begin{equation}\n",
    "i_{e,h} = min(1, max(0, 1 - \\log_{50k}(\\text{IC50}_{e,h})))\n",
    "\\end{equation}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.5 Variables\n",
    "*(Adopted from Pyomo Documentation)*\n",
    "\n",
    "Variables are intended to ultimately be given values by an optimization package. They are declared and optionally bounded, given initial values, and documented using the Pyomo `Var` function. If index sets are given as arguments to this function they are used to index the variable. Other optional directives include:\n",
    "\n",
    "- bounds = A function (or Python object) that gives a (lower,upper) bound pair for the variable\n",
    "- domain = A set that is a super-set of the values the variable can take on.\n",
    "- initialize = A function (or Python object) that gives a starting value for the variable; this is particularly important for non-linear models\n",
    "- within = (synonym for domain)\n",
    "\n",
    "The following code snippet illustrates some aspects of these options by declaring a singleton (i.e. unindexed) variable named `model.LumberJack` that will take on real values between zero and 6 and it initialized to be 1.5:\n",
    "\n",
    "`\n",
    "model.LumberJack = Var(within=NonNegativeReals, bounds=(0,6), initialize=1.5)\n",
    "`\n",
    "\n",
    "\n",
    "For more details see [here](https://pyomo.readthedocs.io/en/latest/pyomo_modeling_components/Variables.html).\n",
    "\n",
    "\n",
    "Task 3.4: OptiTope's Variable definition\n",
    "---\n",
    "Define OptiTope's variables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.6 Objectives\n",
    "*(Adopted from Pyomos Documentation)*\n",
    "\n",
    "An objective is a function of variables that returns a value that an optimization package attempts to maximize or minimize. The `Objective` function in Pyomo declares an objective. Although other mechanisms are possible, this function is typically passed the name of another function that gives the expression. Here is a very simple version of such a function that assumes `model.x` has previously been declared as a `Var`:\n",
    "\n",
    "\n",
    "    def ObjRule(model):\n",
    "        return 2*model.x[1] + 3*model.x[2]\n",
    "    model.g = Objective(rule=ObjRule)\n",
    "\n",
    "\n",
    "It is more common for an objective function to refer to parameters as in this example that assumes that `model.p` has been declared as a `Param` and that `model.x` has been declared with the same index set, while `model.y` has been declared as a singleton:\n",
    "\n",
    "\n",
    "    def profrul(model):\n",
    "        return summation(model.p, model.x) + model.y\n",
    "    model.Obj = Objective(rule=ObjRule, sense=maximize)\n",
    "\n",
    "\n",
    "This example uses the `sense` option to specify maximization. The default `sense` is minimize.\n",
    "\n",
    "Task 3.5: OptiTope's Objective definition\n",
    "---\n",
    "Define OptiTope's objective function\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.7 Constraints\n",
    "*(Adopted from Pyomos Documentation)*\n",
    "\n",
    "Most constraints are specified using equality or inequality expressions that are created using a rule, which is a Python function. For example, if the variable `model.x` has the indexes ‘butter’ and ‘scones’, then this constraint limits the sum over these indexes to be exactly three:\n",
    "\n",
    "    def teaOKrule(model):\n",
    "        return(model.x['butter'] + model.x['scones'] == 3)\n",
    "    model.TeaConst = Constraint(rule=teaOKrule)\n",
    "\n",
    "Instead of expressions involving equality (==) or inequalities (<= or >=), constraints can also be expressed using a 3-tuple if the form (lb, expr, ub) where lb and ub can be None, which is interpreted as lb <= expr <= ub. Variables can appear only in the middle expr. For example, the following two constraint declarations have the same meaning:\n",
    "\n",
    "    model.x = Var()\n",
    "\n",
    "    def aRule(model):\n",
    "       return model.x >= 2\n",
    "    model.Boundx = Constraint(rule=aRule)\n",
    "\n",
    "    def bRule(model):\n",
    "       return (2, model.x, None)\n",
    "    model.boundx = Constraint(rule=bRule)\n",
    "\n",
    "Constraints (and objectives) can be indexed by lists or sets. When the declaration contains lists or sets as arguments, the elements are iteratively passed to the rule function. If there is more than one, then the cross product is sent. For example the following constraint could be interpreted as placing a budget of `i` on the ith item to buy where the cost per item is given by the parameter `model.a`:\n",
    "\n",
    "    model.A = RangeSet(1,10)\n",
    "    model.a = Param(model.A, within=PositiveReals)\n",
    "    model.ToBuy = Var(model.A)\n",
    "    def bud_rule(model, i):\n",
    "        return model.a[i]*model.ToBuy[i] <= i\n",
    "    aBudget = Constraint(model.A, rule=bud_rule)\n",
    "\n",
    "Task 3.6: OptiTope's Constraints\n",
    "---\n",
    "Define OptiTope's constraints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.8 Solving a `ConcreteModels`\n",
    "\n",
    "If you have a ConcreteModel, add these lines at the bottom of your Python script to solve it\n",
    "\n",
    "    opt = pyo.SolverFactory('glpk')\n",
    "    res = opt.solve(model) \n",
    "    \n",
    "    \n",
    "To load the optimal parameters into your model you can use this snippet:\n",
    "\n",
    "    model.solutions.load_from(res)\n",
    "    selected_peptides = [x for x in self.instance.x if 0.9 <= model.x[x].value <= 1.2]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task 3.7: Define now your own objective function for neoepitope selection\n",
    "---\n",
    "Please document and justify your choices as well. You may add additional constraints if naccessary.\n",
    "\n",
    "**NOTE**: You need to turn of the OptiTope's original objective function to solve the problem with your newly defined objective function. You can do so with ```model.<name_of_objective_function>.deactivate()```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Epitope Assembly\n",
    "\n",
    "Once we have selected the top k neoepitopes for vaccination we need to assembly those to the final vaccine construct.\n",
    "Epitope assembly is concerned with finding the best arregement of epitopes to recover all selected peptides after proteasomal cleavage.\n",
    "\n",
    "This problem can be conveniently formulated as [Traveling Salesperson Problem (TSP)](https://en.wikipedia.org/wiki/Travelling_salesman_problem). \n",
    "\n",
    "The TSP tries to find the shortest round trip that visists each city at most once!\n",
    "\n",
    "To translate the assembly problem into a TSP, we interprete the selected epitopes as cities and the negative cleavage likelihood of two adjecent epitopes as the distance between the citities:\n",
    "\n",
    "![img](./img/TSP.png)\n",
    "\n",
    "\n",
    "There are many formulations of the TSP, differing in the number of constraints and in the tightness of the optimality gap of the LP relaxation. A particularly often used formulation was proposed by Miller, Tucker, and Zemlin:\n",
    "\n",
    "\\begin{align}\n",
    "&\\max_{x}\\sum_{a,b\\in E'} w_{a,b}x_{a,b}\\\\\n",
    "\\\\\n",
    "&\\text{subject to}\\\\\n",
    "&\\forall~a \\in E':~~~ \\sum_{b \\in E'} x_{ab} = 1 \\\\\n",
    "&\\forall~a \\in E':~~~ \\sum_{b \\in E'} x_{ba} = 1 \\\\\n",
    "&\\forall~a,b \\in E, a \\neq b:~~~ u_a -u_b+1\\leq (|E'|-1)(1-x_{ab}) \\\\\n",
    "&\\forall~a \\in E: ~2\\leq u_a \\leq |E'|\\\\\n",
    "&u\\in \\mathbb{N_{>0}}\\\\\n",
    "&x\\in \\{0,1\\}\n",
    "\\end{align}\n",
    "\n",
    "with:\n",
    "\n",
    "\\begin{align}\n",
    "& E: \\text{Set of selected epitopes}\\\\\n",
    "& \\epsilon :\\text{Dunmmy epitope} \\\\\n",
    "& E' : \\text{Epitope set with dummy epitope ($E' =E \\cup \\{\\epsilon\\})$}\\\\\n",
    "& w_{ab} : \\text{Edge weight (i.e negative Cleavage likelihood) between epitope $a$ and $b$} \\\\\n",
    "& x_{ab} : \\text{binary decision variable if edge between epitope $a$ and $b$ is selected} \\\\\n",
    "& u_a : \\text{Position of epitope $a$ in string-of-beads ($1\\leq u_a\\leq |E'|$)}\n",
    "\\end{align}\n",
    "\n",
    "Task 4.1: Implement the MTZ-TSP formulation to solve the epitope assembly problem\n",
    "---\n",
    "\n",
    "Use what you have learned to formulate and solve the epitope assembly problem.\n",
    "\n",
    "The following code can be used to predict cleavage likelihoods and to generate the fully-connected acyclic graph:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools as itr\n",
    "from epytope.Core import Peptide\n",
    "from epytope.CleavagePrediction import CleavageSitePredictorFactory\n",
    "\n",
    "pep_tmp = <add here the list of selected neoepitopes as string>\n",
    "pep_tmp.append(\"Dummy\")\n",
    "edge_matrix = {}\n",
    "fragments = {}\n",
    "pred = CleavageSitePredictorFactory(\"proteasmm_c\")\n",
    "\n",
    "for start, stop in itr.combinations(pep_tmp, 2):\n",
    "    if start == \"Dummy\" or stop == \"Dummy\":\n",
    "        edge_matrix[(str(start), str(stop))] = 0\n",
    "        edge_matrix[(str(stop), str(start))] = 0\n",
    "    else:\n",
    "        start_str = str(start)\n",
    "        stop_str = str(stop)\n",
    "        frag = Peptide(start_str+stop_str)\n",
    "        garf = Peptide(stop_str+start_str)\n",
    "\n",
    "        fragments[frag] = (start_str, stop_str)\n",
    "        fragments[garf] = (stop_str, start_str)\n",
    "\n",
    "cleave_pred = pred.predict(fragments.keys())\n",
    "\n",
    "for i in set(cleave_pred.index.get_level_values(0)):\n",
    "    fragment = \"\".join(cleave_pred.loc[i][\"Seq\"])\n",
    "    start, stop = fragments[fragment]\n",
    "    cleav_pos = len(str(start)) - 1\n",
    "    edge_matrix[(start, stop)] = -1.0 * (cleave_pred.loc[(i, len(str(start)) - 1), pred.name])\n",
    "edge_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python [conda env:scanpy] *",
   "language": "python",
   "name": "conda-env-scanpy-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
